### PR TITLE
fix(lora): skip gradient computation for frozen parameters to reduce memory

### PR DESCRIPTION
**Pull request by chen2021673**
## Summary
- Add  needs_input_grad_  tracking to the autograd  Function  base class, populated in  Apply()  before  Forward / SetupContext , so each op knows which inputs actually need gradients
- Linear backward now conditionally skips  grad_input ,  grad_weight , and  grad_bias  computation based on  needs_input_grad_ , avoiding unnecessary GEMM calls and tensor allocations for frozen parameters
-  SetupContext  uses  needs_input_grad_  to selectively save tensors — only saves  input  when  weight  needs grad, and  weight  when  input  needs grad
- Refactor  LinearBackward  kernel signature: replace loose params ( bool transpose ,  int64_t out_features ,  bool bias ) with  LinearMeta  and  LinearGradFlags  structs

## Motivation
During LoRA fine-tuning, the base model weights are frozen ( requires_grad=false ), but the original implementation still allocated and computed full  grad_weight  tensors for every linear layer. This wasted significant GPU memory on gradients that were immediately discarded. With this change, peak GPU memory drops from ~10.7GB to ~7.7GB on the same workload.

### CODE REVIEW

File: infini_train/src/kernels/cuda/linear.cu [hunk 1]
```diff
@@ -323,129 +324,184 @@ __global__ void ReduceColumnsKernel(const T *__restrict__ input, T *__restrict__
 }
 
 std::tuple<std::shared_ptr<Tensor>, std::shared_ptr<Tensor>, std::shared_ptr<Tensor>>
-LinearBackward(const std::shared_ptr<Tensor> &input, const std::shared_ptr<Tensor> &weight, bool transpose,
-               int64_t out_features, const std::shared_ptr<Tensor> &grad_output, const bool bias) {
-    const auto &input_dims = input->Dims();
+LinearBackward(const std::shared_ptr<Tensor> &input, const std::shared_ptr<Tensor> &weight,
+               infini_train::autograd::LinearMeta meta, const std::shared_ptr<Tensor> &grad_output,
+               infini_train::autograd::LinearGradFlags grad_flags) {
```
Comment (kilinchange) [inline infini_train/src/kernels/cuda/linear.cu:324]: kernel 层不应该混入 autograd 级别的内容，

1. meta 直接拆开传参就行；
2. grad_flags 不应该传给 kernel，而应该在 function 级别（infini_train::autograd::Linear::Backward 里）直接拆开调 kernel。（考虑到这是历史遗留，之前写成了融合 kernel，所以这一点这个 pr 不做要求，但后续需要讨论修改）

**Review by kilinchange — CHANGES_REQUESTED** [review]

File: infini_train/src/kernels/cuda/linear.cu (hunk 1 above)
Comment (chen2021673) [reply infini_train/src/kernels/cuda/linear.cu:324]: 拆开调 kernel可以理解，meta 拆开传是什么考虑呢？我觉得作为整体比较清晰，也可以避免频繁修改函数入参

Comment (kilinchange) [reply infini_train/src/kernels/cuda/linear.cu:324]: 主要是从分层解耦的角度考虑。kernel 与 autograd 分层，本质上就是希望后续算子库接入时，只需要依赖 kernel 层定义的接口，而不感知 autograd 层的实现细节。

如果把 `infini_train::autograd::LinearMeta`/`infini_train::autograd::LinearGradFlags` 这类带有 autograd 语义的信息透传到 kernel 层，两层职责边界会变得不够清晰，也会增加后续算子层独立演进的耦合成本。

我理解从整体结构上看，直接传 meta 会更简洁，也能减少函数入参频繁调整；这个方向本身没有问题。但 LinearMeta 当前定义在 autograd 层，从职责划分上不太适合作为 kernel 接口的一部分。

比较理想的方式是：

- 要么在 kernel 这层沉淀计算相关的 meta；
- 要么在 function 层提前拆解，只向 kernel 传递真正计算所需的字段。

考虑到目前还没有特别合适的 kernel 层公共位置，所以建议先拆开传，后续如果类似模式变多，可以再统一抽象一层 kernel 层的公共 meta。

**Comment by kilinchange** [comment]
性能情况：
image

**Comment by kilinchange** [comment]
精度情况：

- lora 存在已知精度波动，后续修复。

image
image

**Review by kilinchange — APPROVED** [review]
